File: project/sup/StudentCollectioniOSJsonRPC/StudentCollectioniOSJsonRPC/Student.swift
/*
 * Copyright 2020 Tim Lindquist,
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Purpose: Example Swift Client for a Java student collection JsonRPC server.
 *
 * Ser423 Mobile Applications
 * see http://pooh.poly.asu.edu/Mobile
 * @author Tim Lindquist Tim.Lindquist@asu.edu
 *         Software Engineering, CIDSE, IAFSE, ASU Poly
 * @version April, 20, 2020
 */

import Foundation

public class Student {
    open var name: String
    open var studentid: Int
    open var takes:[String]=[String]()
    
    init (jsonStr: String){
        self.name = ""
        self.studentid=0
        if let data: Data = jsonStr.data(using: String.Encoding.utf8){
            do{
                let dict = try JSONSerialization.jsonObject(with: data,options:.mutableContainers) as?[String:AnyObject]
                self.name = (dict!["name"] as? String)!
                self.studentid = (dict!["studentid"] as? Int)!
                self.takes = (dict!["takes"] as? [String])!
            } catch {
                print("unable to convert to dictionary")
                
            }
        }
    }
    
    init(dict: [String:Any]){
        self.name = dict["name"] as! String
        self.studentid = dict["studentid"] as! Int
        self.takes = dict["takes"] as! [String]
    }
    
    func toJsonString() -> String {
        var jsonStr = "";
        let dict = ["name": name, "studentid": studentid, "takes":takes] as [String : Any]
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dict, options: JSONSerialization.WritingOptions.prettyPrinted)
            // here "jsonData" is the dictionary encoded in JSON data
            jsonStr = NSString(data: jsonData, encoding: String.Encoding.utf8.rawValue)! as String
        } catch let error as NSError {
            print(error)
        }
        return jsonStr
    }

    func toDict() -> [String:Any] {
        let dict:[String:Any] = ["name": name, "studentid": studentid, "takes":takes] as [String : Any]
        return dict
    }
}

File: project/sup/StudentCollectionJsonRPCServer/src/server/Student.java
package ser423.student.server;
/*
 * Copyright 2020 Tim Lindquist,
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Purpose: Example Java JsonRPC server for collection of students.
 *
 * Ser423 Mobile Applications
 * see http://pooh.poly.asu.edu/Mobile
 * @author Tim Lindquist Tim.Lindquist@asu.edu
 *         Software Engineering, CIDSE, IAFSE, ASU Poly
 * @version February, 2020
 */

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Vector;
import java.util.Arrays;

public class Student {
   public String name;
   public int studentid;
   public Vector<String> takes;

   Student(String name, int studentid, String[] courses){
      this.name = name;
      this.studentid = studentid;
      this.takes = new Vector<String>();
      this.takes.addAll(Arrays.asList(courses));
   }

   Student(String jsonStr){
      try{
         JSONObject jo = new JSONObject(jsonStr);
         name = jo.getString("name");
         studentid = jo.getInt("studentid");
         takes = new Vector<String>();
         JSONArray ja = jo.getJSONArray("takes");
         for (int i=0; i< ja.length(); i++){
            takes.add(ja.getString(i));
         }
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting from json string");
      }
   }

   public Student(JSONObject jsonObj){
      try{
         System.out.println("constructor from json received: "+
                            jsonObj.toString());
         name = jsonObj.getString("name");
         studentid = jsonObj.getInt("studentid");
         takes = new Vector<String>();
         JSONArray ja = jsonObj.getJSONArray("takes");
         for (int i=0; i< ja.length(); i++){
            takes.add(ja.getString(i));
         }
      }catch(Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting from json string");
      }
   }

   public JSONObject toJson(){
      JSONObject jo = new JSONObject();
      try{
         jo.put("name",name);
         jo.put("studentid",studentid);
         jo.put("takes",takes);
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting to json");
      }
      return jo;
   }

   public String toJsonString(){
      String ret = "";
      try{
         ret = this.toJson().toString();
      }catch (Exception ex){
         System.out.println(this.getClass().getSimpleName()+
                            ": error converting to json string");
      }
      return ret;
   }
}
